### PR TITLE
curriculum api

### PR DESCRIPTION
**Pull request by dmitry-bely**

### CODE REVIEW

File: curriculum/build.gradle [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+bootJar {
+    dependsOn asciidoctor
+    from ("${asciidoctor.outputDir}/html5") {
+        into 'static/docs'
+    }
+}
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-actuator'
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
Comment (xpinjection) [inline curriculum/build.gradle]: We don't need JPA at that stage.

File: curriculum/src/main/java/com/kyriba/curriculum/SwaggerConfig.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import springfox.documentation.spring.web.plugins.Docket;
+import springfox.documentation.swagger2.annotations.EnableSwagger2;
+
+
+/**
+ * @author M-DBE
+ */
+@Configuration
+@EnableSwagger2
+@Profile("dev")
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/SwaggerConfig.java]: Why dev profile only?

File: curriculum/src/main/java/com/kyriba/curriculum/SwaggerConfig.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+public class SwaggerConfig
+{
+  @Bean
+  public Docket api()
+  {
+    return new Docket(DocumentationType.SWAGGER_2)
+        .select()
+        .apis(RequestHandlerSelectors.any())
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/SwaggerConfig.java]: It is better to run Swagger only on API and use general apiInfo to describe global things.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
+import java.util.List;
+
+
+/**
+ * @author M-DBE
+ */
+@RestController
+@RequestMapping("/curricula")
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: Strange URL. Usually /api/v1 prefix is used.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+@RequiredArgsConstructor
+public class CurriculumController
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: There is no Swagger documentation for this API.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  @GetMapping(value = "/{id}", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+  @ResponseBody
+  public Curriculum getCurriculumById(@PathVariable("id") long id)
+  {
+    return curriculumService.findCurriculumById(id)
+        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+  }
+
+
+  @GetMapping(value = "/grades/{grade}", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: If result is still Curriculum, then URL is not well designed. At the moment grade is expected as result.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  @ResponseBody
+  public Curriculum getCurriculumByGrade(@Min(1) @Max(11) @PathVariable("grade") int grade)
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: Validation doesn't work, so all these annotations are ignored.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  @ResponseBody
+  public List<Subject> getAllSubjects()
+  {
+    return curriculumService.getAllSubjects();
+  }
+
+
+  @PostMapping(value = "/grades/{grade}", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+  @ResponseBody
+  public BriefCurriculum createCurriculum(@Min(1) @Max(11) @PathVariable("grade") int grade)
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: Again strange URL for Curriculum resource.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  )
+  @ResponseBody
+  public Course removeCourse(@PathVariable("curriculumId") long curriculumId,
+                             @PathVariable("courseId") long courseId)
+  {
+    return curriculumService.removeCourse(curriculumId, courseId);
+  }
+
+
+  @PutMapping(
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: PUT is used to update whole resource or it's part if appropriate URL is used.

File: curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  @ResponseBody
+  public Course updateCourse(@PathVariable("curriculumId") long curriculumId,
+                             @PathVariable("courseId") long courseId,
+                             @RequestParam("lessonCount") int lessonCount)
+  {
+    return curriculumService.updateCourse(curriculumId, courseId, lessonCount);
+  }
+
+
+  @PostMapping(value = "/subjects/{name}")
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/api/CurriculumController.java]: To create subject POST on /subjects URL is used.

File: curriculum/src/main/java/com/kyriba/curriculum/service/CurriculumService.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import com.kyriba.curriculum.domain.dto.Subject;
+
+import java.util.List;
+import java.util.Optional;
+
+
+/**
+ * @author M-DBE
+ */
+public interface CurriculumService
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/service/CurriculumService.java]: We don't need service at that stage.

File: curriculum/src/main/java/com/kyriba/curriculum/service/exception/CurriculumAlreadyExistsException.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright 2000 - 2019. Kyriba Corp. All Rights Reserved.
+ * The content of this file is copyrighted by Kyriba Corporation and can not be
+ * reproduced, distributed, altered or used in any form, in whole or in part.
+ */
+package com.kyriba.curriculum.service.exception;
+
+/**
+ * @author M-DBE
+ */
+public class CurriculumAlreadyExistsException extends CurriculumServiceException
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/service/exception/CurriculumAlreadyExistsException.java]: What curriculum does already exist?

File: curriculum/src/main/java/com/kyriba/curriculum/service/exception/SubjectAlreadyExistsException.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright 2000 - 2019. Kyriba Corp. All Rights Reserved.
+ * The content of this file is copyrighted by Kyriba Corporation and can not be
+ * reproduced, distributed, altered or used in any form, in whole or in part.
+ */
+package com.kyriba.curriculum.service.exception;
+
+/**
+ * @author M-DBE
+ */
+public class SubjectAlreadyExistsException extends RuntimeException
```
Comment (xpinjection) [inline curriculum/src/main/java/com/kyriba/curriculum/service/exception/SubjectAlreadyExistsException.java]: What subject does already exist?

File: curriculum/src/main/java/com/kyriba/curriculum/service/CurriculumService.java (hunk 11 above)
Comment (dmitry-bely) [reply curriculum/src/main/java/com/kyriba/curriculum/service/CurriculumService.java]: I'm not entirely sure what I should with regard to controller. I use service just to separate fake implementation from controller. Should I put fake implementation directly into controller?

Comment (xpinjection) [reply curriculum/src/main/java/com/kyriba/curriculum/service/CurriculumService.java]: At the moment yes. We will focus on implementation in the next modules.

**Pull request merged**